frame/election-providers/src/two_phase/mod.rs

```diff
@@ -0,0 +1,1622 @@
+		/// An account has been rewarded for their signed submission being finalized.
+		Rewarded(<T as frame_system::Config>::AccountId),
+		/// An account has been slashed for submitting an invalid signed submission.
+		Slashed(<T as frame_system::Config>::AccountId),
+		/// The signed phase of the given round has started.
+		SignedPhaseStarted(u32),
+		/// The unsigned phase of the given round has started.
+		UnsignedPhaseStarted(u32),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Submission was too early.
+		EarlySubmission,
+		/// Submission was too weak, score-wise.
+		WeakSubmission,
+		/// The queue was full, and the solution was not better than any of the existing ones.
+		QueueFull,
+		/// The origin failed to pay the deposit.
+		CannotPayDeposit,
+		/// WitnessData is invalid.
+		InvalidWitness,
+	}
+
+	#[pallet::origin]
+	pub struct Origin<T>(PhantomData<T>);
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T>
+	where
+		ExtendedBalance: From<InnerOf<CompactAccuracyOf<T>>>,
+		ExtendedBalance: From<InnerOf<OnChainAccuracyOf<T>>>,
+	{
+		type Call = Call<T>;
+		fn validate_unsigned(source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			if let Call::submit_unsigned(solution, _) = call {
+				// discard solution not coming from the local OCW.
+				match source {
+					TransactionSource::Local | TransactionSource::InBlock => { /* allowed */ }
+					_ => {
+						return InvalidTransaction::Call.into();
+					}
+				}
+
+				let _ = Self::unsigned_pre_dispatch_checks(solution)
+					.map_err(|err| {
+						log!(error, "unsigned transaction validation failed due to {:?}", err);
+						err
+					})
+					.map_err(dispatch_error_to_invalid)
+					.map(Into::into)?;
+
+				ValidTransaction::with_tag_prefix("OffchainElection")
+					// The higher the score[0], the better a solution is.
+					.priority(
+						T::UnsignedPriority::get()
+							.saturating_add(solution.score[0].saturated_into()),
+					)
+					// used to deduplicate unsigned solutions: each validator should produce one
+					// solution per round at most, and solutions are not propagate.
+					.and_provides(solution.round)
+					// transaction should stay in the pool for the duration of the unsigned phase.
+					.longevity(
+						TryInto::<u64>::try_into(T::UnsignedPhase::get())
+							.unwrap_or(crate::two_phase::unsigned::DEFAULT_LONGEVITY),
+					)
+					// We don't propagate this. This can never the validated at a remote node.
+					.propagate(false)
+					.build()
+			} else {
+				InvalidTransaction::Call.into()
+			}
+		}
+
+		fn pre_dispatch(call: &Self::Call) -> Result<(), TransactionValidityError> {
+			if let Call::submit_unsigned(solution, _) = call {
+				Self::unsigned_pre_dispatch_checks(solution)
+					.map_err(dispatch_error_to_invalid)
+					.map_err(Into::into)
+			} else {
+				Err(InvalidTransaction::Call.into())
+			}
+		}
+	}
+
+	#[pallet::type_value]
+	pub fn DefaultForRound() -> u32 {
+		1
```

Comment:
out of curiosity why starting at 1 instead of simply 0

Reply:
I recall that no distinction from `Default` was everything was rather confusing. I will reconsider though, maybe it was only for tests and I can find another workaround.

Reply:
don't bother we can let as it is.